server/pulp/plugins/rsync/publish.py

```diff
@@ -378,30 +364,11 @@ def is_fastforward(self):
         :rtype: bool
         """
         force_full = False
-        for entry in self.predist_history:
-            predistributor_force_full = entry.get("distributor_config", {}).get("force_full",
```

Comment:
I agree I don't see a distributor_config there. However, isn't it a problem?
It seems to me that the intention is correct - to check if there is a force_full configuration on predistributor.
So maybe it makes sense to fix it, aka to get a distributor_id from the RepoPublishResult and then search for its config.

@ipanova, I'm not familiar with rsync distributor so take it with a grain of salt.

Reply:
Since it has been broken for a long time, I wonder if the original intention is really correct. If it is, I'd like to know the reason that rsync publish after a force_full predistributor publish has to be force_full.

Reply:
@emilyzheng , when is force_full publish needed for rsync_distributor?

Reply:
It can be used for fixing content (wrong/missing symlinks for example) in rsync target. It should also be done when publishing with delete=True which needs handling all units instead of incremental units.

Reply:
@emilyzheng as you said force_full fixes wrong/missing symlinks and at a minimum re-creates repodata. So if the predistributor was published in a force full mode you do want to rsync publish in a force full mode as well to bring in new repodata and fixed symlinks

Reply:
I don't think this is needed for the sake of repodata.  If you look at the rsync distributor implementations, if the publish happens at all then they always sync all repodata regardless of whether force_full is used.

I don't think the code can actually know whether a force_full publish is needed.  To me, this is more of an API design question: if caller triggers predistributor with force_full, then can they expect the next rsync distributor to automatically do the same, or is the caller responsible for also specifying that?

If the commit message is right (that this code doesn't work) then the current behavior is that the caller is responsible for deciding whether force_full should be passed to rsync distributor.  The more cautious change is to leave the API behavior alone and not "fix" it to match the original developer's intent, as anyone currently using this must have designed their code to deal with the existing behavior.

Reply:
the last change in this area happened 4 years ago so it is really hard to remember what was the original intent.
I think it makes sense to let the caller explicitly specify `force_full` option as on any other distributor and not make it dependent from predistributor.
In that case we can remove the variable on line 380 and 384

Reply:
Removed.